src/DurableTask.AzureStorage/Partitioning/TablePartitionManager.cs

```diff
@@ -0,0 +1,804 @@
+                                }
+                            }
+                        }
+
+                        //If the lease belongs to a worker that is shutting down, steal it.
+                        if (isOtherWorkerShutDownLease)
+                        {
+                            oldOwner = partition.CurrentOwner!;
+                            this.StealLease(partition);
+                            isStealedLease = true;
+                            response.WaitForPartition = true;
+                        }
+                    }
+
+                }
+
+                // Balancing leases.
+                try
+                {
+                    await this.LeaseBalancer(partitionDistribution, partitions, ownershipLeasesNum, response);
+                }
+                catch (RequestFailedException ex) when (ex.Status == 412)
+                {
+                    throw;
+                }
+                
+
+                return response;
+
+            }
+
+            //This method is for the balance process.
+            //If is, then calculate the average number of leases per worker for balance.
+            //If ownership lease are less than the average number, then steal leases from other workers who have extra leases.
+            //The difference between workers's lease num should not exceed 1.
+            //Therefore, each worker should either have the average number of leases or one more than the average if the total number of leases cannot be evenly divided among them.
+            //Exception will be thrown if the update operation fails due to outdated etagso that worker could re-read the table again to get the latest information.
+            //Any other exceptions will be caught to logs. 
+            public async Task LeaseBalancer(Dictionary<string, List<TableLease>> partitionDistribution, Pageable<TableLease> partitions, int ownershipLeasesNum, ReadTableReponse response)
+            {
+                //Check if there is any other active worker, if not, skip the balance process.
+                if (partitionDistribution.Count == 0)
+                {
+                    return;
+                }
+
+                int averageLeases = (partitions.Count()) / (partitionDistribution.Count + 1);                          
+                if (averageLeases < ownershipLeasesNum)
+                {
+                    // Already have enough leases, and thus no need to steal, just stop the balancing.
+                    return;
+                }
+
+                //Check all the other active workers when my ownership lease number is not larger than the average number.
+                foreach (KeyValuePair<string, List<TableLease>> pair in partitionDistribution)
+                {
+                    int numToSteal = averageLeases - ownershipLeasesNum;
+                    //Already have enough partitions, exit the loop.
+                    if (numToSteal < 0)
```

Comment:
should this be <= 0?
```suggestion
                    if (numToSteal <= 0)
```

Reply:
No. It's a little difficult to explain clearly and please let me know if you have any problems.

So, the balanced goal is that we want the differences between workers no more than 1, which means, 
1. if the number of partitions and the number of workers are dividable, then partitions quota each worker should be the same. For example, 8 partitions with 4 workers, then the distribution should be 2,2,2,2.
2. if the number of partitions and the number of workers are not dividable, then partitions quota each worker couldn't be the same, some worker will have one more quota. For example, 8 partitions with 5 workers, then the distribution should be 1,1,2,2,2.

This setting which you also asked [here](https://github.com/Azure/durabletask/pull/899#discussion_r1235964818) is mainly for the scenario 2.  For example, in this case, the average is 8/5 = 1. And the max partitions a worker could have is 2, which is  (average +1 ). So as long as a worker doesn't reach to the max, it will check other worker's quotas to make sure no one exceed the max. Eg, avoid the distribution 1,1,1,1,4.  That why we have  [here](https://github.com/Azure/durabletask/pull/899#discussion_r1235964818) : `numToSteal` == 0 means the worker already has average number of leases, and `diff` > 1 means the worker it read exceeds the limit of max quota. So it needs to steal from it. 

I think maybe the naming is confusing. I will try to update it to make it more clear.